src/bootstrap/config.rs

```diff
@@ -300,6 +306,13 @@ struct Build {
     configure_args: Option<Vec<String>>,
     local_rebuild: Option<bool>,
     print_step_timings: Option<bool>,
+    stage: Option<u32>,
```

Comment:
The above config shows `"complicated"` but why is it `u32` here?

Reply:
Because it's a stage: either it's not present, or it's an integer.

Reply:
But `"none"` is not an integer right?

Reply:
"none" is meant to represent "not present". If you have a better idea please let me know.

Reply:
Ok so there is a subtlety here you've caught:
- `doc_stage` takes over precedence over `stage` in the config
- _but_, `stage` takes precedence over the _default_ for `doc_stage`.

I'm not sure how to word this without making it more confusing ... maybe just add those lines?

Reply:
Oh, looks like I am totally out of the world. How come `stage` can be `"none"` (which is a string) but `doc-stage` is an integer? I didn't know `stage` have any relationship with `doc-stage`, I thought it was just to name the `stage` `"none"`.

Reply:
That's what I'm trying to say: `"none"` is not a valid value, I'm trying to show the _absence_ of a value which AFAIK isn't possible in toml. Maybe something like this?
```toml
stage = /* no default value */
```

Reply:
> I didn't know stage have any relationship with doc-stage, I thought it was just to name the stage "none"

This is documented on the line above: 

> # The default stage to use for a compile.
> #
> # This is overridden by `doc-stage`, `build-stage`, etc.

Reply:
> That's what I'm trying to say: "none" is not a valid value, I'm trying to show the absence of a value which AFAIK isn't possible in toml. Maybe something like this?

Ah, but putting `"none"` felt like a string. Looks confusing, but why can't we just set a default stage so it could be less confusing? It will be overridden if the more specific ones are set right? Or why not even just remove it?

Reply:
> It will be overridden if the more specific ones are set right? 

https://github.com/rust-lang/rust/pull/76625#discussion_r487408735

There are three main use cases I want to support (https://github.com/rust-lang/rust/issues/76619):
1. people compiling from source, who don't care about all these fine subtleties and just want `--stage 2` for everything. They will be setting `stage = 2` in `config.toml`
2. new contributors, who will use the defaults: different stages for each command
3. frequent contributors (like me), who have strong opinions on what the defaults should be: different stages for each command, but with _different defaults_: `doc_stage = 1`.

Removing `stage` doesn't allow use case 1. Setting a default for stage `stage` makes it take precedence over `doc_stage`, disallowing use case 3.

Reply:
I don't think it is a big ask to just say "if you want the same stage for all the subcommands, modify all of them." There's only 7 commands right now (build/test/doc/run/check/dist/install), and several of these are probably either not important to you or already are at the value you want.

Reply:
Ok, so I should remove `stage = ...` in config.toml altogether? And it will use `doc_stage`, etc. if present, and the hard-coded defaults otherwise?

Reply:
> I don't think it is a big ask to just say "if you want the same stage for all the subcommands, modify all of them." There's only 7 commands right now (build/test/doc/run/check/dist/install), and several of these are probably either not important to you or already are at the value you want.

I do think the tradeoff to be way less confusing a lot better, I don't think everyone needs to use all 7 of them.

Reply:
Ok, I removed `stage` in config.toml.